---
title: Imaobong_R_Assignment 2023
output: html_document
date: "2023-03-10"
---


#Installing and loading packages for this assignment
```{r}
install.packages("tidyverse") 
install.packages("tidyr")
install.packages(naniar)
library(tidyverse) 
library(dplyr)
library(reshape2)
library(naniar)
library(ggplot2)
```
# Upload the required files for this assignment
#fang genotype
```{r}
fang <- read_tsv("https://github.com/EEOB-BioData/BCB546_Spring2023/raw/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")
```
#SNPs position  file
```{r}
snp <- read_tsv("https://github.com/EEOB-BioData/BCB546_Spring2023/raw/main/assignments/UNIX_Assignment/snp_position.txt")
```
#Data inspection for fang genotype file
```{r}
typeof(fang)
#This data is a "list" type of vector
```
```{r}
length(fang)
# This data contains 986 elements
```
```{r}
dim(fang)
#There are 2782 rows and 986 columns
```
```{r}
str(fang)
#provide more details of the data frame
```
```{r}
object.size(fang) 
# size of fang file is 23124584 bytes
```
```{r}
names(fang)
#It shows the names attributed to each of the 986 columns 
```
```{r}
sapply(fang, class)
#Shows class ("Character") of each column in the data frame
```
```{r}
table(fang$Group)
#Provides the numbers in each group
```
#Data inspection for SNPs position file
```{r}
typeof(snp)
#This data is a "list" type of vector
```
```{r}
length(snp)
# This data contains 15 elements
```
```{r}
dim(snp)
#There are 983 rows and 15 columns
```
```{r}
object.size(snp) 
# size of fang file is 359384 bytes
```
```{r}
names(snp)
#It shows the names attributed to each of the 15 columns; 1st column is SNP_ID, and the 15th column is count_gene
```
```{r}
sapply(snp, class)
#Shows class ("Character & numeric") of each column in the data frame
```
```{r}
unique(snp$Chromosome)
# Chromosomes that are unique..."unknown", "multiple", and "integer"
```
#Data processing for Fang data
```{r}
#filter out 3 grps of maize and teosinte before viewing the tables
maize_fang <- filter(fang, Group %in% 
              c('ZMMIL','ZMMLR', 'ZMMMR'))  
table(maize_fang$Group)
```
```{r}
teosinte_fang <- filter(fang, Group %in% 
                 c('ZMPBA', 'ZMPIL', 'ZMPJA'))
table(teosinte_fang$Group)
```
```{r}
#Create a new file with the first 3 columns removed (ID, JG_OTU & group). Then viewed the new tibble to confirmed they are removed
maize_fang1 <- maize_fang[-(1:3)] 
view(maize_fang1)
```
```{r}
teosinte_fang1 <- teosinte_fang[-(1:3)]
view(teosinte_fang1)
```

#Data processing for snp data
```{r}
#Remove unwanted column which is the 2nd column "cdv_marker_id" from original snp file
modified_snp <- snp[-2] 
view(modified_snp) 
```

```{r}
#created a new snp file with (3 columns) containing "SNP_ID, Chromosome, Position". viewed the new file and it looks good.
new_snp <- select(modified_snp, c(1,2,3))
view(new_snp)
```

```{r}
#transposing columns to rows for maize and teosinte, then view the new tibble
maize_trans <- t(maize_fang1)
view(maize_trans)

teosinte_trans <- t(teosinte_fang1)
view(teosinte_trans)
```

```{r}
# merging the new files for maize and teosinte, then view to confirm they are all merged
merged_maize <- cbind(new_snp,maize_trans) %>% remove_rownames()
view(merged_maize)

merged_teosinte <- cbind(new_snp,teosinte_trans) %>% remove_rownames()
view(merged_teosinte)
```

#20 files in total for maize: 
#10 maize files with SNPs ordered based on increasing position
```{r}
#Create a new folder for maize in increasing position
dir.create('./maize_file') 
```

```{r}
#SNPs ordered based on increasing position values and with missing data encoded by this symbol: "?"
mis_maize <- as_tibble(lapply(merged_maize, gsub, pattern = "?", replacement = "?", fixed = TRUE))
maize_chromo <- filter(mis_maize, Chromosome != "unknown" & Chromosome != "multiple") 
for (i in 1:length(unique(maize_chromo$Chromosome))) {
in_maizechro <- maize_chromo %>% filter(Chromosome == i) %>% arrange(Position) 
 write.table(in_maizechro, file =
paste("./maize_file/increased_maize",i), quote = F, sep = "\t")
}
```

#10 maize files with SNPs ordered based on decreasing position
```{r}
#Create a new folder for maize in decreasing position
dir.create('./decreased_maize') 
```

```{r}
#SNPs ordered based on decreasing position values and with missing data encoded by this symbol: "-"

demis_maize <- as_tibble(lapply(merged_maize, gsub, pattern = "?", replacement = "-", fixed = TRUE))
demaize_chromo <- filter(demis_maize, Chromosome != "unknown" & Chromosome != "multiple") 
for (i in 1:length(unique(demaize_chromo$Chromosome))) {
de_maizechro <- demaize_chromo %>% filter(Chromosome == i) %>% arrange(desc(Position)) 
 write.table(de_maizechro, file =
paste("./decreased_maize/decreasing_maize",i), quote = F, sep = "\t")
}
```

#20 files in total for teosinte: 
#10 teosinte files with SNPs ordered based on increasing position
```{r}
#Create a new folder for teosinte in increasing position
dir.create('./teosinte_file') 
```

```{r}
#SNPs ordered based on increasing position values and with missing data encoded by this symbol: "?"

mis_teosinte <- as_tibble(lapply(merged_teosinte, gsub, pattern = "?", replacement = "?", fixed = TRUE))
teosinte_chromo <- filter(mis_teosinte, Chromosome != "unknown" & Chromosome != "multiple") 
for (i in 1:length(unique(teosinte_chromo$Chromosome))) {
in_teosintechro <- teosinte_chromo %>% filter(Chromosome == i) %>% arrange(Position) 
 write.table(in_teosintechro, file =
paste("./teosinte_file/increased_teosinte",i), quote = F, sep = "\t")
}
```

```{r}
#Create a new folder for teosinte in decreasing position
dir.create('./decreased_teosinte') 
```

```{r}
#SNPs ordered based on decreasing position values and with missing data encoded by this symbol: "-"

demis_teosinte <- as_tibble(lapply(merged_teosinte, gsub, pattern = "?", replacement = "-", fixed = TRUE))
deteosinte_chromo <- filter(demis_maize, Chromosome != "unknown" & Chromosome != "multiple") 
for (i in 1:length(unique(teosinte_chromo$Chromosome))) {
de_teosintechro <- deteosinte_chromo %>% filter(Chromosome == i) %>% arrange(desc(Position)) 
 write.table(de_teosintechro, file =
paste("./decreased_teosinte/decreasing_teosinte",i), quote = F, sep = "\t")
}
```

#visualization of Data
```{r}
trans_visc <- fang %>% select(-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% as.data.frame() %>% rownames_to_column (., var = "SNP_ID")

#merged new file for maize and teosinte
merged_trans_visc <- merge(new_snp, trans_visc) %>% filter(., Chromosome != "unknown" & Chromosome != "multiple")
view(merged_trans_visc)
```

#Distribution of SNPs per chromosome
```{r}
#This code "x=as.double()" considers Chromosome as a number fill
#This code "as.factor(as.double())" gets chromosome in the right order
```
```{r}
SNPS_order <- ggplot(merged_trans_visc, aes(x=as.double(Chromosome), fill = as.factor(as.double(Chromosome)))) + geom_bar() +
  scale_x_continuous(breaks = 1:10) + theme_bw() + 
  ggtitle("Distribution of SNPs per chromosome") +
  labs(x = "Chromosome", y = "Total SNPs", fill = "Chromosome")
```

#Distribution of SNPs across chromosome
```{r}
SNPS_diversity <- ggplot(merged_trans_visc, aes(x=
as.numeric(Position))) + geom_density(aes(fill =
as.factor(as.double(Chromosome)))) +
facet_wrap(~ as.factor(as.double(Chromosome)), nrow = 2, ncol = 5) + ggtitle("Distribution of SNPs across chromosome") + theme(axis.text.x=element_text(angle = 90)) +
labs(x = "Position", y = "Density", fill = "Chromosome")
```

#Distribution plot of SNPs on and across Chromosomes in pdf
```{r}
pdf("SNP_Plot.pdf") 
print(SNPS_order) 
print(SNPS_diversity)
dev.off()
```

#Missing SNP data visualization
Still working on the rest of my graphs.

